Extend binding of ConfigurationProvider to support Maps and Nested objects, Added post processing of values in configurationSource

anashaka opened the pull request:
```
Now it is possible to bind with help of ConfigurationProvide.bind("prefix", class) the following interfaces:
```

```
  public interface NestedConfigPojo {
    Integer someSetting();
    ConfigPojo nestedPojo();
  }

  public interface UltraNestedConfigPojo {
    Integer someSetting();
    NestedConfigPojo nestedPojo();
  }

public interface MapPojo {
    Map<String, Integer> map();
  }

public interface MapPojo {
    Map<String, ConfigPojo> map();
  }

```

/**
- Property resolver that process links. Implements {@link PropertiesResolver}.
  *
- Source with properties
- d=1
- a.b.c=$link(d)
  *
- will be resolved to
- d=1
- a.b.c=1
  *
- Source with properties
- d.a=1
- d.b.c=2
- x=$link(d)
  *
- will be resolved tp
- d.a=1
- d.b.c=2
- x.a=1
- x.b.c=2
  *
  */

/**
- Property resolver that process placeholders from system properties. Implements {@link PropertiesResolver}.
  *
- Source with properties
- d=${test}
  *
- where system property test=1
- will be resolved to
- d=1
  */

/**
- Property resolver that process tags in keys. Implements {@link PropertiesResolver}.
- Tags can be used as selector for properties. We can get different sets of properties depends on set of tags provided in constructor.
  *
- Source with properties
  *
- a.$dev1.b=1
- $dev1.c=2
- $dev2.c=22
- $dev1.d.$us=3
- $dev1.d.$eu=33
  *
- where tags=["dev1", "us"]
- will be resolved to
  *
- a.b=1
- c=2
- d=3
  */

alexey-kanashevich committed:
Now it is possible to bind the following interfaces:
```
  public interface NestedConfigPojo {
    Integer someSetting();
    ConfigPojo nestedPojo();
  }

  public interface UltraNestedConfigPojo {
    Integer someSetting();
    NestedConfigPojo nestedPojo();
  }

public interface MapPojo {
    Map<String, Integer> map();
  }

public interface MapPojo {
    Map<String, ConfigPojo> map();
  }

```

alexey-kanashevich committed:
Now it is possible to bind the following interfaces:
```
  public interface NestedConfigPojo {
    Integer someSetting();
    ConfigPojo nestedPojo();
  }

  public interface UltraNestedConfigPojo {
    Integer someSetting();
    NestedConfigPojo nestedPojo();
  }

public interface MapPojo {
    Map<String, Integer> map();
  }

public interface MapPojo {
    Map<String, ConfigPojo> map();
  }

```

alexey-kanashevich committed:
merge

alexey-kanashevich committed:
Now it is possible to bind the following interfaces:
```
  public interface NestedConfigPojo {
    Integer someSetting();
    ConfigPojo nestedPojo();
  }

  public interface UltraNestedConfigPojo {
    Integer someSetting();
    NestedConfigPojo nestedPojo();
  }

public interface MapPojo {
    Map<String, Integer> map();
  }

public interface MapPojo {
    Map<String, ConfigPojo> map();
  }

```

alexey-kanashevich committed:
/**
 * Property resolver that process links. Implements {@link PropertiesResolver}.
 *
 * Source with properties
 * d=1
 * a.b.c=$link(d)
 *
 * will be resolved to
 * d=1
 * a.b.c=1
 *
 * Source with properties
 * d.a=1
 * d.b.c=2
 * x=$link(d)
 *
 * will be resolved tp
 * d.a=1
 * d.b.c=2
 * x.a=1
 * x.b.c=2
 *
 */

/**
 * Property resolver that process placeholders from system properties. Implements {@link PropertiesResolver}.
 *
 * Source with properties
 * d=${test}
 *
 * where system property test=1
 * will be resolved to
 * d=1
 */

/**
 * Property resolver that process tags in keys. Implements {@link PropertiesResolver}.
 * Tags can be used as selector for properties. We can get different sets of properties depends on set of tags provided in constructor.
 *
 * Source with properties
 *
 * a.$dev1.b=1
 * $dev1.c=2
 * $dev2.c=22
 * $dev1.d.$us=3
 * $dev1.d.$eu=33
 *
 * where tags=["dev1", "us"]
 * will be resolved to
 *
 * a.b=1
 * c=2
 * d=3
 */

alexey-kanashevich committed:
Excluded non-string types from resolution

baztian commented:
Nice addition. Why does PlaceholderPropertiesResolver only expand system properties? I would prefer to have a more simple default varialbe expansion without the distinction between system properties and other config values. Would be nice if "other config values" would be taken from all configures sources.

    a=1
    b=${a}
    c=${user.home}

alexey-kanashevich commented:
good catch, however, the idea to get other common properties values through $link() keyword and have them in separate config source set (/links directory with file names matched to property prefix), to have clear understanding of property source for simpler understanding and maintains. And to isolate changed and static part.

alexey-kanashevich committed:
/**
 * Property resolver that process allows retrieve subset of properies started from provided key. And override some of them with new value. Implements {@link PropertiesResolver}.
 *
 * Source with properties
 * d.a=1
 * d.b.c=2
 * x.^extends=d
 * x.a=11
 *
 * will be resolved to
 * d.a=1
 * d.b.c=2
 * x.a=11
 * x.b.c=2
 *
 * Similar functionality can be achived with {@link LinkPropertiesResolver}. The main difference that linking achieved through property key and allows override.
 */

alexey-kanashevich committed:
Added to PlaceholderResolver ability to resolve properties from source .

anashaka commented:
@baztian Done. Now values for placeholder searched also in "other config values" (from all configures sources)
Also added support of recursive resolve and several placeholder in one string.
Also implemented ^extends keyword.

alexey-kanashevich committed:
Prevent cast exception for non string types

alexey-kanashevich committed:
Prevent cast exception for non string types

anashaka commented:
@norbertpotocki could you have a look at this PR

norbertpotocki commented:
Hey @anashaka. Let me start by saying that I appreciate your contributions to cfg4j. Well done! As for this pull request I'm a bit hesitant about merging it. I like the nested binding part of it. It's definitely needed and I already started working on it some time ago. The custom syntax for configuration values and keys is another story though. Here's a few observations I have. Let me know what you think.

I've seen attempts at applying the DRY rule to configuration sets at several companies I worked for. Sadly it never worked. The crux of the problem is how people interact with configurations.

* configuration is more often read than written and reading configuration that contains $link(), aliases, etc. is surprisingly hard
* configuration sets are usually small and easy to maintain, so it's very tempting to simply copy-paste parts of it
* you want configuration to be easy to understand even by people who are not familiar with a configuration library that parses it

That said I'm simply not sure if extra overhead of supporting this feature is worth it. Let me know what your thought on this are.

gitmoneyman commented:
Hi @norbertpotocki. Agree and don't agree with you. 
- I agree that each tool should be used right.
- That configuration should be simple and understandable and short.
- we can remove $link keyword it duplicates functionality of ^extends.

From my perspective and experience:
- I successfully applied DSL for configuration last 8 years. It was template engine with generator of properties files first and groovy after that (groovy much better, we build risk platform on the base of it). There are simple rules (lessons learned):
- Configuration should have single entry point 
- if we introduce new keywords(DSL) it is ok if it is less than 7 and meaning of them can be derived from previous experience (spring, programming language, etc) - suppose that learning curve is (10 minutes in worst case)
- If we have imports (links, extends) it should be only one level nesting (import allowed only in root)
- it should be clear where to find imported config (can be solved with convention or with extending link with location)
- imported config should be bounded to separate domain (auth, database(connectionPool,mongo), bus, zipkin)
- From perspective of support resolvers based on properties from sources. If everything ok with properties everything ok with resolvers. Also we already have them in production. I can support you with support)).

I have some examples.
If we have component that participate in 
- environments (local, dev, test, uat, staging, prod) 6
- separate products (only config is different) we have 6.
- in regions of deployment for some of environments (eu, us)
For one such unlucky component I need to have 72 configurations (6 envs * 6 products*2 regions).
Lets assume I've made refactoring in this component and added several configuration properties. Depression waits me in this case. Copy past cannot help here. DRY rule can.Also during such copy past I can make a lot of mistakes.
 I can divide configs to small logical portions and link them in one place together 
To achieve this I need link and filter functionality.
In case of refactoring I will change only one part.
The same story for configuration of external libraries (zipkin, hystryx, etc). If I use them in all components it is quite difficult to update to new version if config changed.
Second example is spring. Spring context it is also configuration. And it have placeholders and profiles. 
I have more but it is not blog post

Also I am going opensource CLI for extended version of configPusher and some docs

alexey-kanashevich commented:
Hi @norbertpotocki. Agree and don't agree with you.

- I agree that each tool should be used right.
- That configuration should be simple and understandable and short.
- we can remove $link keyword it duplicates functionality of ^extends.

From my perspective and experience:

- I successfully applied DSL for configuration last 8 years. It was template engine with generator of properties files first and groovy after that (groovy much better, we build risk platform on the base of it). 
- There are simple rules (lessons learned):
- Configuration should have single entry point
- if we introduce new keywords(DSL) it is ok if it is less than 7 and meaning of them can be derived from previous experience (spring, programming language, etc) - suppose that learning curve is (10 minutes in worst case)
- If we have imports (links, extends) it should be only one level nesting (import allowed only in root)
- it should be clear where to find imported config (can be solved with convention or with extending link with location)
- imported config should be bounded to separate domain (auth, database(connectionPool,mongo), bus, zipkin)
- From perspective of support resolvers based on properties from sources. If everything ok with properties everything ok with resolvers. Also we already have them in production. I can support you with support)).
 

I have some examples.

If we have component that participate in:

- environments (local, dev, test, uat, staging, prod) 6
- separate products (only config is different) we have 6.
- in regions of deployment for some of environments (eu, us) For one such unlucky component I need 

to have 72 configurations (6 envs * 6 products*2 regions). Lets assume I've made refactoring in this component and added several configuration properties. Depression waits me in this case. Copy past cannot help here. DRY rule can.Also during such copy past I can make a lot of mistakes. I can divide configs to small logical portions and link them in one place together To achieve this I need link and filter functionality. In case of refactoring I will change only one part. The same story for configuration of external libraries (zipkin, hystryx, etc). If I use them in all components it is quite difficult to update to new version if config changed. 

Second example is spring. Spring context it is also configuration. And it have placeholders and profiles.
 I have more but it is not blog post.
Also I am going opensource CLI for extended version of configPusher and some docs

jdiazcano commented:
Is this being worked on? I would really like to see nested binding!

remmeier commented:
+1 it would be great to see this things in the repo

alexey-kanashevich committed:
fixed bug with unprocessed placeholder if it has digits

remmeier commented:
FYI: related  to this I also opened up a pull request for https://github.com/cfg4j/cfg4j/pull/193. This would further extend the object binding with annotations that allow the declaration of default values. 

```
  @DefaultValue("13")
  public int intValue();
```

those value can be access by hooking AnnotationConfigurationSource into cfg4j.

alexey-kanashevich committed:
add support of enum keys to map binding

alexey-kanashevich committed:
revert build gradle

rocketraman commented:
+1 Some kind of nested binding is definitely needed. Perhaps @anashaka you could split up this pull request into one for nested binding, and then a separate one that builds on top for links/aliases. That at least gets the nested binding capability in, and may make the smaller pulls more palatable for @norbertpotocki ?

remmeier commented:
various PRs are open, maybe we should consider forking the project off somewhere and go ahead with the various requests.

gitmoneyman commented:
You can use my fork, it is stable and used in production, maybe we can setup publish artifacts to public repo for convinience with name cfg4j-on-steroids

anashaka commented:
You can use my fork it is stable and works in production. Maybe I need to setup publishing of artifacts to public repo. Proposed name cfg4j-on-steroids

anashaka committed:
ISSUE-209:for ConsulConfigurationSource if some errors happens in the reload process once, the initialized flag will keep false all the way, which will lead to the getConfiguration function throw IllegalStateException

anashaka committed:
ISSUE-209: Replace double check with try to init if not initialized

anashaka committed:
Merge branch 'master' into master

anashaka committed:
ISSUE-209: fix compile errors

roded commented:
I have no strong opinion regarding linking, but resolving values from system properties is definitely a must.
Please take this into consideration.

fabriziocucci commented:
I don't know how many people are still waiting for this but I ended up implementing this in [my library](https://github.com/fabriziocucci/yacl4j). Both nested configurations and placeholder resolution are there.

roded commented:
@fabriziocucci the experimental status is a bit concerning :)
@anashaka, did you get a chance to publish your work to a public repo? Would this be possible?

fabriziocucci commented:
@roded I've been using it for quite some time now and I hardly see it changing, I should remove that section actually! :) Anyway, if you happen to give it a try, feedback are more than welcome.

lucarosellini commented:
@anashaka is your fork available in any public repo?

kant111 commented:
Can't wait for this! When can I expect this on maven repo?  I am currently using 4.4.0

remmeier commented:
may let us all settle on a new name and repo and continue there as there is no progress here. Potentially do some minor refactoring one the way.

kant111 commented:
@remmeier Any reason why this PR is not merged yet especially when all checks have passed?

remmeier commented:
no owner around

fabriziocucci commented:
@kant111, as suggested in on one of my previous comments, if you really need this feature and you can't wait till this PR goes through, give [yacl4j](https://github.com/fabriziocucci/yacl4j/) a try!

vjeko commented:
It would be great if this library had support for nested interfaces.

anashaka committed:
Can process non string values (numbers) from system properties

anashaka committed:
Extends keyword allows override

anashaka committed:
Added optional support

anashaka committed:
set priorities for label more concrete wins.

anashaka committed:
Migrate to consul client of spring-boot

jsicheng commented:
Nested support would be great. CFG4J uses [type-parser](https://github.com/drapostolos/type-parser/wiki/#supported-arraycollectionmap-types) for deserialization. Some simple objects (eg: Maps) can be expressed in string format.

Example here: [java](https://github.com/cfg4j/cfg4j-sample-apps/blob/master/classpath-bind/src/main/java/org/cfg4j/sample/ReksioConfig.java#L19) [yaml](https://github.com/cfg4j/cfg4j-sample-apps/blob/master/classpath-bind/src/main/resources/application.yaml#L29)
